### PR TITLE
fix windows build

### PR DESCRIPTION
**Pull request by houqp**
# Description

Fix windows build, which was blocking us from releasing a windows
python client. Added mac and window build to the build matrix.

# Related Issue(s)

Also move rename temp file clean up logic into put_obj method in
preparation for moving the temp file creation logic into optimistic
delta commit loop. See https://github.com/delta-io/delta-rs/issues/135.

### CODE REVIEW

**Comment by houqp** [comment]
cc @mosyp and @xianwill since this is related to #135.

File: rust/src/storage/file/mod.rs
Comment (houqp) [inline rust/src/storage/file/mod.rs:80]: switched to async rename

File: rust/src/delta.rs
Comment (houqp) [inline rust/src/delta.rs:301]: turns out we can't use `join_path` to build regex query in a cross-platform way because the separator `/` from windows is a regex escape character.

File: rust/src/storage/file/rename.rs
Comment (rtyler) [inline rust/src/storage/file/rename.rs:23]: Linux specific code in the unix-specific implementation? I'm quite confused by why these libc functions need to be mapped in? There's a high likelihood this could break on BSDs, non-glibc implementations.

(I haven't tested it of course, so you may know something I don't)

Isn't there a crate that abstracts this behavior which doesn't require the `unsafe` below?  E.g. https://docs.rs/crate/atomicwrites/0.3.0


Of course, if you've already gone down this path and this is the best possible option, then so be it, I would appreciate some commentary in the code explaining why "this must be"

File: rust/src/storage/file/rename.rs
Comment (rtyler) [inline rust/src/storage/file/rename.rs:66]: Reading the manpage for this function this little bit jumped out to me

> RENAME_NOREPLACE **requires support from the underlying filesystem.**

Emphasis mine of course. I don't have a good alternative suggestion, but just would like to express my worry that there couplings to underlying system implementation details that we cannot clearly communicate to users here.

File: rust/src/storage/file/rename.rs
Comment (houqp) [reply rust/src/storage/file/rename.rs:23]: platform specific atomic rename needs to be implemented per platform unfortunately, so BSD and non-glibc users will just need to send PRs for the platforms they want to use, just like windows.

I looked into atomicwrites before, their implementation is not ideal. It requires 3 system calls to achieve what can be done in 1 system call, which is the implementation we have here. I also don't think their implementation is correct, because their implementation started with a hardlink function call, which based on the std rust doc, it doesn't error out if destination path already exists.

The unsafe code we have here is scoped to only system call invocation, which is basically what happens when you call `std::fs::rename`. So even though atomicwrites has less unsafe code on the surface, it's just hidden under the std lib implementation.

File: rust/src/storage/file/rename.rs
Comment (houqp) [reply rust/src/storage/file/rename.rs:66]: good call, added rust doc for filesystem back to clarify that.

**Review by rtyler — APPROVED** [review]
Looks good to me, I think @fvaleye or @mosyp should take a look before we merge

File: rust/src/storage/file/rename.rs
Comment (mosyp) [reply rust/src/storage/file/rename.rs:23]: @rtyler 
> why these libc functions need to be mapped in

Because they're missing in libc crate and PR for introducing them is still in open

Comment (mosyp) [reply rust/src/storage/file/rename.rs:23]: well https://github.com/rust-lang/libc/pull/2116#issuecomment-801511926 seems like they're merged this a couple of days ago

**Review by mosyp — APPROVED** [review]
lgtm!

File: rust/src/storage/file/rename.rs
Comment (houqp) [reply rust/src/storage/file/rename.rs:23]: ha nice, filed https://github.com/delta-io/delta-rs/issues/143 as follow up

**Pull request merged**
